Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,379 @@
+  preimage f := f.φ PUnit.unit
+
+instance : (of C).Full :=
+  fullyFaithfulOf.full
+
+instance : (of C).Faithful :=
+  fullyFaithfulOf.faithful
+
+
+section Coproduct
+
+variable (𝒜 : Type w) (f : 𝒜 → FormalCoproduct.{w} C) (t X : FormalCoproduct.{w} C)
+
+def cofan : Cofan f :=
+  Cofan.mk ⟨(i : 𝒜) × (f i).I, fun p ↦ (f p.1).obj p.2⟩
+    fun i ↦ ⟨fun x ↦ ⟨i, x⟩, fun x ↦ 𝟙 ((f i).obj x)⟩
+
+section simp_lemmas
+
+variable {𝒜 f}
+
+theorem cofan_inj (i : 𝒜) : (cofan 𝒜 f).inj i = ⟨fun x ↦ ⟨i, x⟩, fun x ↦ 𝟙 ((f i).obj x)⟩ := rfl
+-- JH: this is probably a bad theorem? At least not good for simp
+
+@[simp] lemma cofan_inj_f_fst (i : 𝒜) (x) : (((cofan 𝒜 f).inj i).f x).1 = i := rfl
+
+@[simp] lemma cofan_inj_f_snd (i : 𝒜) (x) : (((cofan 𝒜 f).inj i).f x).2 = x := rfl
+
+@[simp] lemma cofan_inj_φ (i : 𝒜) (x) : ((cofan 𝒜 f).inj i).φ x = 𝟙 ((f i).obj x) := rfl
+
+end simp_lemmas
+
+@[simps!] def cofanHomEquiv :
+    ((cofan 𝒜 f).pt ⟶ t) ≃ ((i : 𝒜) → (f i ⟶ t)) where
+  toFun m i := (cofan 𝒜 f).inj i ≫ m
+  invFun s := ⟨fun p ↦ (s p.1).f p.2, fun p ↦ (s p.1).φ p.2⟩
+  left_inv m := hom_ext rfl (fun ⟨i, x⟩ ↦ by simp [cofan_inj])
+  right_inv p := by ext <;> simp
+
+@[simps!] def isColimitCofan : IsColimit (cofan 𝒜 f) :=
+  mkCofanColimit (cofan 𝒜 f) (fun t ↦ (cofanHomEquiv _ _ _).symm t.inj)
+    (fun t i ↦ congrFun ((cofanHomEquiv _ _ _).right_inv t.inj) i)
+    (fun _ _ h ↦ (Equiv.eq_symm_apply _).2 (funext h))
+
+instance : HasCoproducts.{w} (FormalCoproduct.{w} C) :=
+  hasCoproducts_of_colimit_cofans _ (isColimitCofan _)
+
+noncomputable def coproductIsoCofan : ∐ f ≅ (cofan 𝒜 f).pt :=
+  colimit.isoColimitCocone ⟨_, isColimitCofan _ _⟩
+
+variable {𝒜 f} in
+@[reassoc (attr := simp)] lemma ι_comp_coproductIsoCofan (i) :
+    Sigma.ι f i ≫ (coproductIsoCofan 𝒜 f).hom = (cofan 𝒜 f).inj i :=
+  colimit.isoColimitCocone_ι_hom _ _
+
+def toFun (X : FormalCoproduct.{w} C) : X.I → FormalCoproduct.{w} C :=
+  (of C).obj ∘ X.obj
+
+def coproductCoconeIsoSelf : (cofan X.I X.toFun).pt ≅ X :=
+  isoOfComponents (Equiv.sigmaPUnit X.I) fun i ↦ Iso.refl (X.obj i.fst)
+
+@[reassoc (attr := simp)]
+lemma inj_comp_coproductCoconeIsoSelf (i : X.I) :
+    (cofan X.I X.toFun).inj i ≫ (coproductCoconeIsoSelf X).hom = ofHom.mk i (𝟙 (X.obj i)) :=
+  hom_ext rfl (fun i => by simp; rfl)
+
+@[simps!] noncomputable def coproductIsoSelf :
+    ∐ X.toFun ≅ X :=
+  coproductIsoCofan _ _ ≪≫ coproductCoconeIsoSelf X
+
+@[reassoc (attr := simp)] lemma ι_comp_coproductIsoSelf (i : X.I) :
+    Sigma.ι _ i ≫ (coproductIsoSelf X).hom = ofHom.mk i (𝟙 (X.obj i)) := by
+  simp [coproductIsoSelf]
+
+end Coproduct
+
+
+noncomputable
+instance [HasTerminal C] (X : FormalCoproduct.{w} C) : Unique (X ⟶ (of C).obj (⊤_ C)) :=
+  ⟨⟨⟨fun _ ↦ PUnit.unit, fun _ ↦ terminal.from _⟩⟩,
+  fun j ↦ hom_ext (funext fun _ ↦ rfl) (by aesop)⟩
+
+instance [HasTerminal C] : HasTerminal (FormalCoproduct.{w} C) :=
+  (IsTerminal.ofUnique <| (of C).obj (⊤_ C)).hasTerminal
```

Comment:
Maybe this can be rephrased as `PreservesTerminal (of C)`? This way, you can also get `Nonempty (IsTerminal ((of C).obj x))` for any terminal object of `C` and not just the abstract one?

Reply:
ditto

Reply:
Refer to Zulip: [#mathlib4 > Preserves terminal @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Preserves.20terminal/near/526089633)

Since we currently do not have a consistent way to work with terminal objects, I've opted for the following approach:
```lean
section Terminal

def isTerminalOf (T : C) (ht : IsTerminal T) : IsTerminal ((of C).obj T) :=
  IsTerminal.ofUniqueHom (fun _ ↦ ⟨fun _ ↦ PUnit.unit, fun _ ↦ ht.from _⟩)
    (fun _ _ ↦ hom_ext (funext fun _ ↦ rfl) (fun _ ↦ ht.hom_ext _ _))

instance [HasTerminal C] : HasTerminal (FormalCoproduct.{w} C) :=
  (isTerminalOf (⊤_ C) terminalIsTerminal).hasTerminal

end Terminal
```

Reply:
> we currently do not have a consistent way to work with terminal objects

What do you mean by this? The intended design is to have a statement in terms of `IsTerminal` to work with any terminal object, and a version with `HasTerminal` to work with the one chosen by choice, as you've done above

Reply:
Apologies, I've hallucinated a `PreservesTerminal` type class, in my first comment which may have caused confusion.
To me, the statement with `IsTerminal ((of C).obj x)`  when `IsTerminal x` + registering a `HasTerminal` instance when `HasTerminal C` is a good way to state things.